### ISSUE TITLE
Некритичная проблема с вызовом TI_GetPortfolio()

### ISSUE THREAD

**Issue by Mefodyev** [ErhoSen/gas-tinkoff-trades#15]
Вызов TI_GetPortfolio() периодически возвращает
`Exception: Request failed for https://invest-public-api.tinkoff.ru returned code 429 (строка 271).` 

Не могу уловить порядок воспроизведения ошибки, возникает от случая к случаю. В круглых скобках ссылаюсь на номер счета из другого листа в виде `=TI_GetPortfolio('Счета'!A2)`. Если писать в скобках сам номер, не поможет. Все подгрузится, только если из формулы убрать "=" и вставить заново. Причем, с обычным счетом портфолио подгружается сразу, а с ИИС падает чаще. Остальные функции работают корректно.

Хочется чтоб при открытии документа все погружалось сразу, без вмешательства в ячейки. Возможно, я что-то настроил не до конца. Подскажите пож. нормально ли это и как избавиться от ошибки.

**Comment by HiHat** [comment]
Думаю, что это ошибка API и ее надо регистрировать тут: https://github.com/Tinkoff/investAPI/issues

**Comment by HiHat** [comment]
Видимо неправильно создали параметр в настройках проекта.
Примерно так должно  быть:
![Снимок экрана 2023-06-30 в 22 23 42](https://github.com/ErhoSen/gas-tinkoff-trades/assets/2690010/f3a9708d-c3dc-4da4-8a0a-ebacda605672)

**Comment by Mefodyev** [comment]
@HiHat хочу поднять тот же вопрос. Ошибку не регал, т.к. она на время уходила. Сейчас кол-во инструментов в портфеле увеличилось и она вернулась. По моим наблюдениям, суть во времени выполнения запроса (~30 секунд).  С портфелями, где инструментов мало, или запрос идет менее 30 секунд, все в порядке:
![image](https://github.com/erhosen/gas-tinkoff-trades/assets/40695873/00445034-4754-4e95-aaf1-14a874aa1b68)

Но если их много и запрос длится больше 30 секунд, то выполняется с ошибкой:
<img width="1332" alt="Снимок экрана 2024-01-02 в 19 21 56" src="https://github.com/erhosen/gas-tinkoff-trades/assets/40695873/c1c0ed10-99ce-4682-abcc-3b14b0a8b258">

И причина с предпоследним невыполненным шагом (не дождался конца):
<img width="727" alt="Снимок экрана 2024-01-02 в 22 10 38" src="https://github.com/erhosen/gas-tinkoff-trades/assets/40695873/6ec109b9-a362-4e23-b5a9-0458a759d55a">

В документации их API не лазил, не знаю правильно ли описанное мной поведение. Уточняю на всякий случай, ошибка только(?) на их стороне и о ней тоже надо сообщать сюда? https://github.com/Tinkoff/investAPI/issues

**Comment by PozhidaevGM** [comment]
Я когда-то делал для своих целей форк без оформления тут, где решал проблему за счет кэширования инструментов (большой портфель не может обработаться, т.к. там на каждый элемент заново идёт обращение в тинёк о получении его характеристик). Т.е. решить можно только за счёт того, что если "сегодня" функцию еще не вызывал, то нужно будет запустить функцию несколько раз, чтобы кэш заполнился и 30 секунд хватало.
Вечером выложу кусок, может кто-то из вас интегрирует (или сам закодит)

**Comment by HiHat** [comment]
> Я когда-то делал для своих целей форк без оформления тут, где решал проблему за счет кэширования инструментов (большой портфель не может обработаться, т.к. там на каждый элемент заново идёт обращение в тинёк о получении его характеристик). Т.е. решить можно только за счёт того, что если "сегодня" функцию еще не вызывал, то нужно будет запустить функцию несколько раз, чтобы кэш заполнился и 30 секунд хватало. Вечером выложу кусок, может кто-то из вас интегрирует (или сам закодит)

Будем благодарны,  если поделитесь примером!
Почитал немного об ограничениях обхода лимита на время выполнения за 30 секунд и способах обхода, но пока не придумал как лучше реализовать!

**Comment by HiHat** [comment]
> В документации их API не лазил, не знаю правильно ли описанное мной поведение. Уточняю на всякий случай, ошибка только(?) на их стороне и о ней тоже надо сообщать сюда? https://github.com/Tinkoff/investAPI/issues

Нет, похоже при больших портфелях происходит слишком много запросов к API и срабатывает ограничение на 23.5 секунды, которое скрипт пытается подождать, но его ожидание прерывается средой GAS, т.к. на бесплатном тарифе функция должна выполнятся не более чем за 30 секунд.
Надеюсь найдём способ как оптимизировать скрипт и обойти ограничения API и среды GAS!

**Comment by HiHat** [comment]
При подготовке информации по портфелю, по каждому инструменту через API запрашивается его тикер.
Можно сделать быстрый вариант вывода портфеля,  где будут только FIGI инструментов! :)

**Comment by Mefodyev** [comment]
> При подготовке информации по портфелю, по каждому инструменту через API запрашивается его тикер. Можно сделать быстрый вариант вывода портфеля, где будут только FIGI инструментов! :)

Нет, не вариант, смысл вывода всего портфеля был в возможности построить детализированную сводную таблицу со всей нужной инфой. Сама по себе фига никакого бизнес-смысла не несёт.
Либо (правда проблему это не решит), можно сделать урезанный вариант getportfolio без кол-ва бумаг, ср. цены покупки, ст-ти покупки, нкд, и валюта три раза почему-то возвращается.

**Comment by PozhidaevGM** [comment]
> Я когда-то делал для своих целей форк без оформления тут, где решал проблему за счет кэширования инструментов (большой портфель не может обработаться, т.к. там на каждый элемент заново идёт обращение в тинёк о получении его характеристик). Т.е. решить можно только за счёт того, что если "сегодня" функцию еще не вызывал, то нужно будет запустить функцию несколько раз, чтобы кэш заполнился и 30 секунд хватало. Вечером выложу кусок, может кто-то из вас интегрирует (или сам закодит)

HiHat уже сделал кэширование инструментов по фигам, значит выхода два:
1. Костыль: запускать функцию несколько раз, пока все фиги не закэшируются; Как просрочится кэш - заново.
2. Выгрузить на дополнительный лист фиги-имя, и "ВПРом" их подставлять потом уже на листе вне GAS (закоммитил сейчас изменения для вывода такой таблицы)

ЗЫЖ Обратите внимание, что фиги есть далеко не у всех инструментов, и надо наверное переходить на Uid

**Comment by HiHat** [comment]
В общем пока реализовано только кеширование получения имен инструментов по FIGI.
На больших портфелях может работать не с первого раза (пока кэш не наполниться).
Позже, возможно, придумаем, как обойти ограничение в 30 секунд на выполнение функций.
